src/http/matcher/mod.rs

```diff
@@ -65,9 +68,30 @@ pub enum HttpMatcherKind {
     ///
     /// [`SocketAddr`]: std::net::SocketAddr
     Socket(SocketMatcher),
+    /// A custom matcher that implements [`crate::service::Matcher`].
+    // Custom(Box<dyn crate::service::Matcher<State, Request<Body>>>), TODO: Does not satisfy Clone, unless inner field satisfies Clone.
+    Custom(Arc<dyn crate::service::Matcher<State, Request<Body>>>),
+}
+
+impl<State, Body> fmt::Debug for HttpMatcherKind<State, Body> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // match self {
+        //     Self::All(inner) => f.debug_tuple("All").field(inner).finish(),
+        //     Self::Method(inner) => f.debug_tuple("Method").field(inner).finish(),
+        //     Self::Path(inner) => f.debug_tuple("Path").field(inner).finish(),
+        //     Self::Domain(inner) => f.debug_tuple("Domain").field(inner).finish(),
+        //     Self::Version(inner) => f.debug_tuple("Version").field(inner).finish(),
+        //     Self::Any(inner) => f.debug_tuple("Any").field(inner).finish(),
+        //     Self::Uri(inner) => f.debug_tuple("Uri").field(inner).finish(),
+        //     Self::Header(inner) => f.debug_tuple("Header").field(inner).finish(),
+        //     Self::Socket(inner) => f.debug_tuple("Socket").field(inner).finish(),
+        //     Self::Custom(_) => f.debug_tuple("Custom").finish(),
```

Comment:
And then you can add the field here as well :)

Reply:
Even for the inner field `Arc<dyn crate::service::Matcher<State, Request<Body>>>`?

I left this out since I did not think we wanted to add a `Debug` trait bound for this, as I imagine it would be a closure in most cases? I also see for [AllowOrigin::Predicate](https://github.com/plabayo/rama/blob/main/src/http/layer/cors/allow_origin.rs#L113) the inner field is omitted.

Reply:
I would do that as otherwise there is no way to see the inner field. And anyway you are expected to implement debug for everything, while at same time ensuring you hide (manually) sensitive info. E.g. for something containing a password you would implement debug manually to not show password, etc..

But if you feel strongly against then I'm also fine for it, we can always introduce it later in a new breaking release.

Reply:
Feels a bit much. For now you can just do your original plan, output it without the actual inner field exposed.

we can revisit this in future if there are complaints about it. There will still be plenty of breaking releases, so plenty of opportunities to put this straight if it has to be.